### ISSUE TITLE
Reduce CPU load in idle mode

### ISSUE THREAD

**Issue by hofheinz** [vifm/vifm#72]
While idling (i.e., main screen, no keys pressed, waiting for user input), vifm uses about 0.5% cpu time on my machine. (I tested the current master branch and the current debian package, and the numbers are similar on my desktop and laptop computers, both with up-to-date hardware and under a default debian install.)

This is not dramatic but may be somewhat problematic in a low-end environment with several vifm instances open, and might also impact power usage on a laptop (because the cpu wakes up all the time, see below).

I did an strace, and I think this is related to the way main_loop.c waits for a keypress (read_char() in main_loop.c). With the default settings (timeoutlen=1000), there are about 130 syscalls per second. Most of them are "poll" requests that wait for a keystroke (15ms timeout), but there are also other ones (re-reading /etc/passwd and /etc/group about once per second, and some curses stuff).

I've played around with the constants T, IPC_F in read_char(), and with the configuration option timeoutlen a bit, and the load is almost zero for large T and timeoutlen, and small IPC_F. (As far as I can tell, this means "do nothing unless a key is actually pressed".)

So here's my **question** (or, feature request maybe): would it be possible to add a possibility (configuration option?) to adjust the tradeoff between cpu usage and always displaying up-to-date information? (I imagine this might be potentially helpful in low-end or laptop environments.)

Sorry for the long post. Also, I really appreciate the work that you put into this, xaizek, it's a great and very useful file manager!

**Comment by xaizek** [comment]
I've noticed this too, but it didn't bother me that much.

That loop does quite a lot of things:

1. Processes keyboard input.
2. Checks whether directory was changed since last check.
3. Checks for IPC messages, which is actually *disabled by default* and it's my bad that `IPC_F` isn't set to `1` in this case. And even if it's enabled, only one instance that is in "server" mode need higher rate of checks. Thanks, this is a small, but still improvement.
4. Checks status of background jobs. This should be OK to move out of `for` loop in `read_char()`.
5. Updates interface, this shouldn't require any extra work unless terminal was resized or something like that.

Looks like `read_char()` is not the best name for the function : -).

One thing I wanted to add is an option to configure directory update rate, which should set the `T` constant. Maybe with these three changes (IPC, background jobs, polling rate) it will behave much better. If not, we can go from there.

Thanks for investigating this and all the details. I'll add a comment when I commit these changes (soon).

**Comment by hofheinz** [comment]
Thanks for the very quick and helpful answer! Let me know if I can help. (I don't feel confident enough with the code to supply a patch, but I'd be happy to test.)

**Comment by xaizek** [comment]
@hofheinz, please see `reduce-cpu-usage-while-idle` branch if you want to give it a try. Also found a couple of more things that could be cached.

The new option to configure that `T` is called `mintimeoutlen`. In my tests with `timeoutlen=1000 mintimeoutlen=1000`, looks like I get only three system calls per second, with `mintimeoutlen=150` it's around 30 calls.

**Comment by hofheinz** [comment]
Sounds great! I'll test this later today.

**Comment by hofheinz** [comment]
I can confirm that the option works as expected. I tested this with an empty setup and with my own. In both cases, the numbers I got from strace were a little different: for (timeoutlen,mintimeoutlen)=(1000,1000), I got about 6 syscalls, and for (1000,150), I got about 20. (For (10000,10000), it's about 0.5 syscalls, so this can really be adjusted to extreme settings.)

What I did was: strace -o log vifm, and then count the lines in log (after and before the init and exit phases). The syscalls came always in triples of two "stats" (for the directories in the panes), and one "select" call (I suppose this one comes from reading from stdin). I think I have yet to understand what exactly you did there (and where the other calls went) by looking at the diff, but this looks exactly like what's minimally necessary. Looks very good!

**Comment by xaizek** [comment]
Great, thanks for the testing!

Duplication of system calls with such settings is my bad, should be fixed with the top-most commit (`strace tt -o log ./vifm` displayed real timestamps, so this time it was really easy to check the rate and understand the cause).

Other calls were almost extra calls, it's either iteration of the main loop, which could be skipped, or querying system or internal information that changes rarely.  I just added caching of some data and update it when there is real need for this.

I'll add some cosmetic changes for code nearby code I touched and then merge it into `master`.
